test/e2e/directory_permissions_test.go

```diff
@@ -0,0 +1,171 @@
+	"gopkg.in/yaml.v2"
+)
+
+type filePerms map[string]os.FileMode
+
+func TestDirectoryPermissions(t *testing.T) {
+	env := BuildEnv(t)
+	vendir := Vendir{t, env.BinaryPath, Logger{}}
+
+	tCases := map[string]struct {
+		updateConfig  func(cfg *config.Config)
+		expectedPerms filePerms
+	}{
+		"no permissions defined": {
+			expectedPerms: filePerms{
+				"dir-0": 0700, filepath.Join("dir-0", "subdir-0-0"): 0700, filepath.Join("dir-0", "subdir-0-1"): 0700,
+				"dir-1": 0700, filepath.Join("dir-1", "subdir-1-0"): 0700, filepath.Join("dir-1", "subdir-1-1"): 0700,
+			},
+		},
+		"outer dir permissions can be configured": {
+			updateConfig: func(c *config.Config) {
+				c.Directories[0].Permission = p(0744)
+			},
+			expectedPerms: filePerms{
+				"dir-0": 0744, filepath.Join("dir-0", "subdir-0-0"): 0744, filepath.Join("dir-0", "subdir-0-1"): 0744,
+				"dir-1": 0700, filepath.Join("dir-1", "subdir-1-0"): 0700, filepath.Join("dir-1", "subdir-1-1"): 0700,
+			},
+		},
+		"inner dir permissions can be configured": {
+			updateConfig: func(c *config.Config) {
+				c.Directories[0].Contents[0].Permission = p(0755)
+				c.Directories[0].Contents[1].Permission = p(0744)
+			},
+			expectedPerms: filePerms{
+				"dir-0": 0700, filepath.Join("dir-0", "subdir-0-0"): 0755, filepath.Join("dir-0", "subdir-0-1"): 0744,
+				"dir-1": 0700, filepath.Join("dir-1", "subdir-1-0"): 0700, filepath.Join("dir-1", "subdir-1-1"): 0700,
+			},
+		},
+		"blocking write or execute in (sub)dirs still works": {
+			updateConfig: func(c *config.Config) {
+				c.Directories[0].Permission = p(0100) // we still need exec permissions here, so that we can stat its subdirectories
+				c.Directories[0].Contents[0].Permission = p(0000)
+				c.Directories[0].Contents[1].Permission = p(0004)
+			},
+			expectedPerms: filePerms{
+				"dir-0": 0100, filepath.Join("dir-0", "subdir-0-0"): 0000, filepath.Join("dir-0", "subdir-0-1"): 0004,
+				"dir-1": 0700, filepath.Join("dir-1", "subdir-1-0"): 0700, filepath.Join("dir-1", "subdir-1-1"): 0700,
+			},
+		},
+	}
+
+	for tName, tCase := range tCases {
+		t.Run(tName, func(t *testing.T) {
+			cfg := defaultConfig()
+
+			if u := tCase.updateConfig; u != nil {
+				u(cfg)
+			}
+
+			actualPerms := runAndGetActualPerms(t, vendir, *cfg)
+			tCase.expectedPerms.validate(t, actualPerms)
+		})
+	}
+}
+
+func p(p os.FileMode) *os.FileMode { return &p }
+
+func (expected filePerms) validate(t *testing.T, actual filePerms) {
+	for path, expectedPerms := range expected {
+		actualPerms, ok := actual[path]
+		assert.True(t, ok, "no actual permissions for path %s found", path)
+		assert.Equal(t, expectedPerms, actualPerms, "expected permissions for '%s' to be '%s', but got '%s'", path, expectedPerms, actualPerms)
+	}
+}
+
+func writeConfigFile(t *testing.T, tmpDir string, config config.Config) {
+	vendirYML, err := os.Create(filepath.Join(tmpDir, "vendir.yml"))
+	assert.NoError(t, err, "creating vendir.yml")
```

Comment:
I think we should use require in this place as well as in other parts. The main difference between `require` and `assert` is that assert allows the these to continue, but in situations like this we should not continue with the execution of the test. (I did not highlight all the places where this change can be done)
```suggestion
	require.NoError(t, err, "creating vendir.yml")
```